recommenders/models/sasrec/util.py

```diff
@@ -0,0 +1,309 @@
+            continue
+
+        seq = np.zeros([maxlen], dtype=np.int32)
+        idx = maxlen - 1
+        for i in reversed(train[u]):
+            seq[idx] = i
+            idx -= 1
+            if idx == -1:
+                break
+
+        rated = set(train[u])
+        rated.add(0)
+        item_idx = [valid[u][0]]
+        for _ in range(num_neg_test):
+            t = np.random.randint(1, itemnum + 1)
+            while t in rated:
+                t = np.random.randint(1, itemnum + 1)
+            item_idx.append(t)
+
+        inputs = {}
+        inputs["user"] = np.expand_dims(np.array([u]), axis=-1)
+        inputs["input_seq"] = np.array([seq])
+        inputs["candidate"] = np.array([item_idx])
+
+        # predictions = -model.predict(sess, [u], [seq], item_idx)
+        predictions = -1.0 * model.predict(inputs)
+        predictions = np.array(predictions)
+        predictions = predictions[0]
+
+        rank = predictions.argsort().argsort()[0]
+
+        valid_user += 1
+
+        if rank < 10:
+            NDCG += 1 / np.log2(rank + 2)
+            HT += 1
+
+    return NDCG / valid_user, HT / valid_user
+
+
+def create_combined_dataset(u, seq, pos, neg, seq_max_len):
+    """
+    function to create model inputs from sampled batch data.
+    This function is used only during training.
+    """
+    inputs = {}
+    seq = tf.keras.preprocessing.sequence.pad_sequences(
+        seq, padding="pre", truncating="pre", maxlen=seq_max_len
+    )
+    pos = tf.keras.preprocessing.sequence.pad_sequences(
+        pos, padding="pre", truncating="pre", maxlen=seq_max_len
+    )
+    neg = tf.keras.preprocessing.sequence.pad_sequences(
+        neg, padding="pre", truncating="pre", maxlen=seq_max_len
+    )
+
+    inputs["users"] = np.expand_dims(np.array(u), axis=-1)
+    inputs["input_seq"] = seq
+    inputs["positive"] = pos
+    inputs["negative"] = neg
+
+    target = np.concatenate(
+        [
+            np.repeat(1, seq.shape[0] * seq.shape[1]),
+            np.repeat(0, seq.shape[0] * seq.shape[1]),
+        ],
+        axis=0,
+    )
+    target = np.expand_dims(target, axis=-1)
+    return inputs, target
+
+
+def train(model, dataset, sampler, **kwargs):
+    """
+    High level function for model training
+    """
+    num_epochs = kwargs.get("num_epochs", 10)
+    maxlen = kwargs.get("maxlen", 100)
+    num_neg_test = kwargs.get("num_neg_test", 100)
+    batch_size = kwargs.get("batch_size", 128)
+    lr = kwargs.get("learning_rate", 0.001)
+
+    [user_train, user_valid, user_test, usernum, itemnum] = dataset
+    num_steps = int(len(user_train) / batch_size)
+
+    optimizer = tf.keras.optimizers.Adam(
+        learning_rate=lr, beta_1=0.9, beta_2=0.999, epsilon=1e-7
+    )
+
+    loss_function = model.loss_function
+
+    train_loss = tf.keras.metrics.Mean(name="train_loss")
+
+    train_step_signature = [
+        {
+            "users": tf.TensorSpec(shape=(None, 1), dtype=tf.int64),
```

Comment:
Minor point here, just for consistency, in most or all the other ML models we have the same pattern for training: `model.fit` see https://github.com/microsoft/recommenders/blob/main/examples/00_quick_start/lstur_MIND.ipynb instead of a wrapper function whose input is a model and dataset. Same with the prediction. See for example when we predict for ranking: https://github.com/microsoft/recommenders/blob/main/recommenders/models/sar/sar_singlenode.py#L450 and when we predict for rating: https://github.com/microsoft/recommenders/blob/main/recommenders/models/sar/sar_singlenode.py#L482